include/vkgcDefs.h

```diff
@@ -749,4 +761,208 @@ class IPipelineDumper {
 
 };
 
+class EntryHandle;
+
+/// 128-bit hash compatible structure
+struct HashId
+{
+  union
+  {
+    uint64_t ddwords[2]; ///< Output hash in dwords.
+    uint32_t dwords[4];  ///< Output hash in dwords.
+    uint8_t  bytes[16];  ///< Output hash in bytes.
+  };
+};
+
+// =====================================================================================================================
+// Shader Cache interfaces, that client needs to inherit and implement it.
+class ICache
+{
+public:
+  virtual ~ICache() {}
+
+  /// \brief Obtain a cache entry for the \p hash.
+  ///
+  /// The caller receives reference-counted ownership of the returned handle, if any, and must
+  /// eventually call \ref PutEntry to release it.
+  ///
+  /// Valid handles are always non-null.
+  ///
+  /// \param hash the hash key for the cache entry
+  /// \param allocateOnMiss if true, a new cache entry will be allocated when none is found
+  /// \param pHandle will be filled with a handle to the cache entry
+  /// \return success code, possible values:
+  ///   * Success: an existing, ready entry was found
+  ///   * NotReady: an existing entry was found, but it is not ready yet because another thread
+  ///               is working on filling it
+  ///   * NotFound: no existing entry was found. If \p allocateOnMiss is set, a new entry was
+  ///               allocated and the caller must populate it via SetValue
+  ///   * ErrorXxx: some internal error has occurred, no handle is returned
+  virtual Result GetEntry(HashId hash,
+                          bool allocateOnMiss,
+                          EntryHandle *pHandle) = 0;
+
+  /// \brief Release ownership of a handle to a cache entry.
+  ///
+  /// If the caller is responsible for populating the cache entry, it is an error to call this
+  /// method without first calling SetValue.
+  /// It can put Entry multiple times if Entry is empty.
+  ///
+  /// \param rawHandle the handle to cache entry to be released
+  virtual void PutEntry(void *rawHandle) = 0;
```

Comment:
It's a single handle, why needs change to rawHandle**s**?
The function name is a pair with GetEntry. We have refcount concept behind them.

Reply:
Sorry, what I meant is that you are passing around a `void *` and it's easy to pass some unrelated object by accident. Would it be possible to introduce a type for `rawHandle` so that you can only use pointers of correct type?

Reply:
Is Get and Put a standard naming in refcounting? Objective C uses retain and release. To me Get and Release sounds much more descriptive.